#### Sử dụng pandas để đọc dữ liệu

In [1]:
# Set up, nhập thư viện

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score


In [2]:
# Nhập dữ liệu

df = pd.read_csv('Case_study_CarPrice_Assignment.csv')

In [3]:
# xem 10 dòng ngẫu nhiên

df.sample(10)

car_ID  symboling                   CarName fueltype aspiration  \
154     155          0  toyota corolla 1600 (sw)      gas        std   
84       85          3      mitsubishi mirage g4      gas      turbo   
64       65          0                 mazda 626      gas        std   
159     160          0            toyota corolla   diesel        std   
97       98          1               nissan note      gas        std   
82       83          3      mitsubishi outlander      gas      turbo   
176     177         -1            toyota corolla      gas        std   
181     182         -1            toyouta tercel      gas        std   
204     205         -1               volvo 264gl      gas      turbo   
118     119          1         plymouth fury iii      gas        std   

    doornumber    carbody drivewheel enginelocation  wheelbase  ...  \
154       four      wagon        4wd          front       95.7  ...   
84         two  hatchback        fwd          front       95.9  ...   
64        four  hatchback        fwd          front       98.8  ...   
159       four  hatchback        fwd          front       95.7  ...   
97        four      wagon        fwd          front       94.5  ...   
82         two  hatchback        fwd          front       95.9  ...   
176       four      sedan        fwd          front      102.4  ...   
181       four      wagon        rwd          front      104.5  ...   
204       four      sedan        rwd          front      109.1  ...   
118        two  hatchback        fwd          front       93.7  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio horsepower  \
154          92        2bbl       3.05    3.03              9.0         62   
84          156        spdi       3.59    3.86              7.0        145   
64          122        2bbl       3.39    3.39              8.6         84   
159         110         idi       3.27    3.35             22.5         56   
97           97        2bbl       3.15    3.29              9.4         69   
82          156        spdi       3.58    3.86              7.0        145   
176         122        mpfi       3.31    3.54              8.7         92   
181         161        mpfi       3.27    3.35              9.2        156   
204         141        mpfi       3.78    3.15              9.5        114   
118          90        2bbl       2.97    3.23              9.4         68   

     peakrpm citympg  highwaympg    price  
154     4800      27          32   7898.0  
84      5000      19          24  14489.0  
64      4800      26          32  11245.0  
159     4500      38          47   7788.0  
97      5200      31          37   7999.0  
82      5000      19          24  12629.0  
176     4200      27          32  10898.0  
181     5200      19          24  15750.0  
204     5400      19          25  22625.0  
118     5500      37          41   5572.0  

[10 rows x 26 columns]

In [4]:
# car_ID : Mã xe
# symboling: Mức độ an toàn (giá trị trong khoảng từ 2 tới 3, trong đó giá trị -2: tuyệt
# đối an toàn, 3: có rủi ro)
# carName: Tên xe
# fueltype: Loại nhiên liệu
# aspiration: loại hút (turbo hoặc std)
# doornumber: Số lượng cửa
# carbody: Loại thân xe (sedan, wagon, hatchback)
# drivewheel: Loại bánh xe
# enginelocation: Vị trí động cơ
# wheelbase: Chiều dài cơ sở
# carlength: Chiều dài xe
# carwidth: Chiều rộng xe
# carheight: Chiều cao xe
# curbweight: Trọng lượng của xe khi không có người hoặc hành lý
# enginetype: Loại động cơ
# cylindernumber: Số lượng xi lanh đặt trong xe
# enginesize: Kích thước động cơ
# fuelsystem: Hệ thống nhiên liệu của xe ô tô
# boreratio: Tỉ số hành trình
# stroke: số kỳ hoăc kích thước bên trong động cơ
# compressionratio: Tỉ số nén của ôtô
# horsepower: Mã lực
# peakrpm: Tốc độ động cơ cao nhất
# citympg: số dặm đi được cho 1 galon (~4.5 l) nhiên liệu trong thành phố
# highwaympg: số dặm đi được cho 1 galon (~4.5 l) nhiên liệu trên đường cao tốc
# price: Giá xe


In [5]:
# xem DL có bao nhiêu cột, dòng

df.shape

(205, 26)

## Xem xét kiểu dữ liệu của các thuộc tính, thực hiện chuyển đổi về đúng kiểu 

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

##### Nhận xét: 
dữ liệu không có dữ liệu thiếu

In [7]:
df.columns

Index(['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [8]:
# Thay thế các giá trị của cột cylindernumber và doornumber từ kiểu 
# dữ liệu obj sang int

In [9]:
df.cylindernumber

0      four
1      four
2       six
3      four
4      five
       ... 
200    four
201    four
202     six
203     six
204    four
Name: cylindernumber, Length: 205, dtype: object

In [10]:
df['cylindernumber'].unique()

array(['four', 'six', 'five', 'three', 'twelve', 'two', 'eight'],
      dtype=object)

In [11]:
df.doornumber.unique()

array(['two', 'four'], dtype=object)

In [12]:
df['doornumber'] = df['doornumber'].replace({'two':2, 'four':4})
df['cylindernumber'] = df['cylindernumber'].replace({'four':4, 'six':6, 'five':5, 'three':3, 'twelve':12, 'two':2, 'eight':8})

df

car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
203     204         -1                 volvo 246   diesel      turbo   
204     205         -1               volvo 264gl      gas      turbo   

     doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0             2  convertible        rwd          front       88.6  ...   
1             2  convertible        rwd          front       88.6  ...   
2             2    hatchback        rwd          front       94.5  ...   
3             4        sedan        fwd          front       99.8  ...   
4             4        sedan        4wd          front       99.4  ...   
..          ...          ...        ...            ...        ...  ...   
200           4        sedan        rwd          front      109.1  ...   
201           4        sedan        rwd          front      109.1  ...   
202           4        sedan        rwd          front      109.1  ...   
203           4        sedan        rwd          front      109.1  ...   
204           4        sedan        rwd          front      109.1  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio  horsepower  \
0           130        mpfi       3.47    2.68              9.0         111   
1           130        mpfi       3.47    2.68              9.0         111   
2           152        mpfi       2.68    3.47              9.0         154   
3           109        mpfi       3.19    3.40             10.0         102   
4           136        mpfi       3.19    3.40              8.0         115   
..          ...         ...        ...     ...              ...         ...   
200         141        mpfi       3.78    3.15              9.5         114   
201         141        mpfi       3.78    3.15              8.7         160   
202         173        mpfi       3.58    2.87              8.8         134   
203         145         idi       3.01    3.40             23.0         106   
204         141        mpfi       3.78    3.15              9.5         114   

     peakrpm citympg  highwaympg    price  
0       5000      21          27  13495.0  
1       5000      21          27  16500.0  
2       5000      19          26  16500.0  
3       5500      24          30  13950.0  
4       5500      18          22  17450.0  
..       ...     ...         ...      ...  
200     5400      23          28  16845.0  
201     5300      19          25  19045.0  
202     5500      18          23  21485.0  
203     4800      26          27  22470.0  
204     5400      19          25  22625.0  

[205 rows x 26 columns]

In [13]:
df.cylindernumber

0      4
1      4
2      6
3      4
4      5
      ..
200    4
201    4
202    6
203    6
204    4
Name: cylindernumber, Length: 205, dtype: int64

In [14]:
# Xem thông kê mô tả của dữ liệu

df.describe()

car_ID   symboling  doornumber   wheelbase   carlength    carwidth  \
count  205.000000  205.000000  205.000000  205.000000  205.000000  205.000000   
mean   103.000000    0.834146    3.121951   98.756585  174.049268   65.907805   
std     59.322565    1.245307    0.994966    6.021776   12.337289    2.145204   
min      1.000000   -2.000000    2.000000   86.600000  141.100000   60.300000   
25%     52.000000    0.000000    2.000000   94.500000  166.300000   64.100000   
50%    103.000000    1.000000    4.000000   97.000000  173.200000   65.500000   
75%    154.000000    2.000000    4.000000  102.400000  183.100000   66.900000   
max    205.000000    3.000000    4.000000  120.900000  208.100000   72.300000   

        carheight   curbweight  cylindernumber  enginesize   boreratio  \
count  205.000000   205.000000      205.000000  205.000000  205.000000   
mean    53.724878  2555.565854        4.380488  126.907317    3.329756   
std      2.443522   520.680204        1.080854   41.642693    0.270844   
min     47.800000  1488.000000        2.000000   61.000000    2.540000   
25%     52.000000  2145.000000        4.000000   97.000000    3.150000   
50%     54.100000  2414.000000        4.000000  120.000000    3.310000   
75%     55.500000  2935.000000        4.000000  141.000000    3.580000   
max     59.800000  4066.000000       12.000000  326.000000    3.940000   

           stroke  compressionratio  horsepower      peakrpm     citympg  \
count  205.000000        205.000000  205.000000   205.000000  205.000000   
mean     3.255415         10.142537  104.117073  5125.121951   25.219512   
std      0.313597          3.972040   39.544167   476.985643    6.542142   
min      2.070000          7.000000   48.000000  4150.000000   13.000000   
25%      3.110000          8.600000   70.000000  4800.000000   19.000000   
50%      3.290000          9.000000   95.000000  5200.000000   24.000000   
75%      3.410000          9.400000  116.000000  5500.000000   30.000000   
max      4.170000         23.000000  288.000000  6600.000000   49.000000   

       highwaympg         price  
count  205.000000    205.000000  
mean    30.751220  13276.710571  
std      6.886443   7988.852332  
min     16.000000   5118.000000  
25%     25.000000   7788.000000  
50%     30.000000  10295.000000  
75%     34.000000  16503.000000  
max     54.000000  45400.000000

### Tìm mối liên hệ giữa hãng xe và tên xe, phát hiện và sửa sai dữ liệu

In [15]:
df.CarName

0            alfa-romero giulia
1           alfa-romero stelvio
2      alfa-romero Quadrifoglio
3                   audi 100 ls
4                    audi 100ls
                 ...           
200             volvo 145e (sw)
201                 volvo 144ea
202                 volvo 244dl
203                   volvo 246
204                 volvo 264gl
Name: CarName, Length: 205, dtype: object

In [16]:
# tách cột BrandName 

df['BrandName'] = df['CarName'].str.split(' ').str[0]
df

car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
203     204         -1                 volvo 246   diesel      turbo   
204     205         -1               volvo 264gl      gas      turbo   

     doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0             2  convertible        rwd          front       88.6  ...   
1             2  convertible        rwd          front       88.6  ...   
2             2    hatchback        rwd          front       94.5  ...   
3             4        sedan        fwd          front       99.8  ...   
4             4        sedan        4wd          front       99.4  ...   
..          ...          ...        ...            ...        ...  ...   
200           4        sedan        rwd          front      109.1  ...   
201           4        sedan        rwd          front      109.1  ...   
202           4        sedan        rwd          front      109.1  ...   
203           4        sedan        rwd          front      109.1  ...   
204           4        sedan        rwd          front      109.1  ...   

     fuelsystem  boreratio  stroke  compressionratio horsepower  peakrpm  \
0          mpfi       3.47    2.68               9.0        111     5000   
1          mpfi       3.47    2.68               9.0        111     5000   
2          mpfi       2.68    3.47               9.0        154     5000   
3          mpfi       3.19    3.40              10.0        102     5500   
4          mpfi       3.19    3.40               8.0        115     5500   
..          ...        ...     ...               ...        ...      ...   
200        mpfi       3.78    3.15               9.5        114     5400   
201        mpfi       3.78    3.15               8.7        160     5300   
202        mpfi       3.58    2.87               8.8        134     5500   
203         idi       3.01    3.40              23.0        106     4800   
204        mpfi       3.78    3.15               9.5        114     5400   

     citympg highwaympg    price    BrandName  
0         21         27  13495.0  alfa-romero  
1         21         27  16500.0  alfa-romero  
2         19         26  16500.0  alfa-romero  
3         24         30  13950.0         audi  
4         18         22  17450.0         audi  
..       ...        ...      ...          ...  
200       23         28  16845.0        volvo  
201       19         25  19045.0        volvo  
202       18         23  21485.0        volvo  
203       26         27  22470.0        volvo  
204       19         25  22625.0        volvo  

[205 rows x 27 columns]

In [17]:
df['BrandName'].unique()

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'maxda', 'mazda', 'buick', 'mercury',
       'mitsubishi', 'Nissan', 'nissan', 'peugeot', 'plymouth', 'porsche',
       'porcshce', 'renault', 'saab', 'subaru', 'toyota', 'toyouta',
       'vokswagen', 'volkswagen', 'vw', 'volvo'], dtype=object)

## Tiến hành phân tích mối liên hệ của từng thuộc tính với price bằng các biểu đồ mô tả tần số, tần suất, các phép kiểm định

Nhận xét các thuộc tính gây ảnh hưởng tới giá


##### Hướng làm:

1.trực quan hoá

2.vẽ biểu đồ scatter/line plot với biến định lượng

bar plot với biến định tính

--> thấy biến nào có tương quan thì thử chạy kiểm định giả thuyết tương quan

In [18]:
# tách dữ liệu thành 2 phần: gồm các biến định lượng và phần định tính

# df1 gồm các cột có kiểu float hoặc int
df1 = df[['cylindernumber', 'doornumber', 'car_ID', 'symboling', 'wheelbase', 'carlength', 'carwidth', 'carheight', 'curbweight', 'enginesize',
'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg', 'price']]
df1.head()

cylindernumber  doornumber  car_ID  symboling  wheelbase  carlength  \
0               4           2       1          3       88.6      168.8   
1               4           2       2          3       88.6      168.8   
2               6           2       3          1       94.5      171.2   
3               4           4       4          2       99.8      176.6   
4               5           4       5          2       99.4      176.6   

   carwidth  carheight  curbweight  enginesize  boreratio  stroke  \
0      64.1       48.8        2548         130       3.47    2.68   
1      64.1       48.8        2548         130       3.47    2.68   
2      65.5       52.4        2823         152       2.68    3.47   
3      66.2       54.3        2337         109       3.19    3.40   
4      66.4       54.3        2824         136       3.19    3.40   

   compressionratio  horsepower  peakrpm  citympg  highwaympg    price  
0               9.0         111     5000       21          27  13495.0  
1               9.0         111     5000       21          27  16500.0  
2               9.0         154     5000       19          26  16500.0  
3              10.0         102     5500       24          30  13950.0  
4               8.0         115     5500       18          22  17450.0

In [19]:
# df2 gồm các cột định lượng và cột price
 
df2 = df[['CarName', 'fueltype', 'aspiration', 'carbody', 'drivewheel', 'enginelocation', 'enginetype', 'fuelsystem', 'price']]
df2.head()

CarName fueltype aspiration      carbody drivewheel  \
0        alfa-romero giulia      gas        std  convertible        rwd   
1       alfa-romero stelvio      gas        std  convertible        rwd   
2  alfa-romero Quadrifoglio      gas        std    hatchback        rwd   
3               audi 100 ls      gas        std        sedan        fwd   
4                audi 100ls      gas        std        sedan        4wd   

  enginelocation enginetype fuelsystem    price  
0          front       dohc       mpfi  13495.0  
1          front       dohc       mpfi  16500.0  
2          front       ohcv       mpfi  16500.0  
3          front        ohc       mpfi  13950.0  
4          front        ohc       mpfi  17450.0

In [20]:
# bảng df1 xem mối quan hệ giữa các biến độc lập với biến dự đoán price

df1.corr()

cylindernumber  doornumber    car_ID  symboling  wheelbase  \
cylindernumber          1.000000   -0.016009 -0.094493  -0.113129   0.339507   
doornumber             -0.016009    1.000000  0.190352  -0.664073   0.447357   
car_ID                 -0.094493    0.190352  1.000000  -0.151621   0.129729   
symboling              -0.113129   -0.664073 -0.151621   1.000000  -0.531954   
wheelbase               0.339507    0.447357  0.129729  -0.531954   1.000000   
carlength               0.430672    0.398568  0.170636  -0.357612   0.874587   
carwidth                0.545007    0.207168  0.052387  -0.232919   0.795144   
carheight              -0.013995    0.552208  0.255960  -0.541038   0.589435   
curbweight              0.609727    0.197379  0.071962  -0.227691   0.776386   
enginesize              0.846031    0.020742 -0.033930  -0.105790   0.569329   
boreratio               0.231399    0.119258  0.260064  -0.130051   0.488750   
stroke                  0.008210   -0.011082 -0.160824  -0.008735   0.160959   
compressionratio       -0.020002    0.177888  0.150276  -0.178515   0.249786   
horsepower              0.692016   -0.126947 -0.015006   0.070873   0.353294   
peakrpm                -0.124172   -0.247668 -0.203789   0.273606  -0.360469   
citympg                -0.445837   -0.012417  0.015940  -0.035823  -0.470414   
highwaympg             -0.466666   -0.036330  0.011255   0.034606  -0.544082   
price                   0.718305    0.031835 -0.109093  -0.079978   0.577816   

                  carlength  carwidth  carheight  curbweight  enginesize  \
cylindernumber     0.430672  0.545007  -0.013995    0.609727    0.846031   
doornumber         0.398568  0.207168   0.552208    0.197379    0.020742   
car_ID             0.170636  0.052387   0.255960    0.071962   -0.033930   
symboling         -0.357612 -0.232919  -0.541038   -0.227691   -0.105790   
wheelbase          0.874587  0.795144   0.589435    0.776386    0.569329   
carlength          1.000000  0.841118   0.491029    0.877728    0.683360   
carwidth           0.841118  1.000000   0.279210    0.867032    0.735433   
carheight          0.491029  0.279210   1.000000    0.295572    0.067149   
curbweight         0.877728  0.867032   0.295572    1.000000    0.850594   
enginesize         0.683360  0.735433   0.067149    0.850594    1.000000   
boreratio          0.606454  0.559150   0.171071    0.648480    0.583774   
stroke             0.129533  0.182942  -0.055307    0.168790    0.203129   
compressionratio   0.158414  0.181129   0.261214    0.151362    0.028971   
horsepower         0.552623  0.640732  -0.108802    0.750739    0.809769   
peakrpm           -0.287242 -0.220012  -0.320411   -0.266243   -0.244660   
citympg           -0.670909 -0.642704  -0.048640   -0.757414   -0.653658   
highwaympg        -0.704662 -0.677218  -0.107358   -0.797465   -0.677470   
price              0.682920  0.759325   0.119336    0.835305    0.874145   

                  boreratio    stroke  compressionratio  horsepower   peakrpm  \
cylindernumber     0.231399  0.008210         -0.020002    0.692016 -0.124172   
doornumber         0.119258 -0.011082          0.177888   -0.126947 -0.247668   
car_ID             0.260064 -0.160824          0.150276   -0.015006 -0.203789   
symboling         -0.130051 -0.008735         -0.178515    0.070873  0.273606   
wheelbase          0.488750  0.160959          0.249786    0.353294 -0.360469   
carlength          0.606454  0.129533          0.158414    0.552623 -0.287242   
carwidth           0.559150  0.182942          0.181129    0.640732 -0.220012   
carheight          0.171071 -0.055307          0.261214   -0.108802 -0.320411   
curbweight         0.648480  0.168790          0.151362    0.750739 -0.266243   
enginesize         0.583774  0.203129          0.028971    0.809769 -0.244660   
boreratio          1.000000 -0.055909          0.005197    0.573677 -0.254976   
stroke            -0.055909  1.000000          0.186110    0.080940 -0.067964   

# Nhận xét:
từ bảng trên ta có thể nhận thấy một số biến có mối quan hệ với biến 'price' là các biến

cylindernumber, wheelbase,carlength, carwidth, curbweight, enginesize, boreratio, highwaympg	

--> chạy kiểm định tương quan -> chạy 8 lần kiểm định

In [21]:
# Nhập thư viện
from scipy import stats

In [22]:
# Kiểm định Giữa cylindernumber và price có tương quan với nhau hay không

df1_1 = df1[['cylindernumber', 'price']]

print(stats.pearsonr(df1_1['cylindernumber'], df1_1['price']))

PearsonRResult(statistic=0.7183049020002832, pvalue=8.028863372444613e-34)


### nhận xét: pvalue=8.028863372444613e-34 < 0.05 
--> bác bỏ H0 -> tức là 2 thuộc tính này (cylindernumber và price) có tương quan

In [23]:
# Kiểm định Giữa wheelbase và price có tương quan với nhau hay không

df1_2 = df1[['wheelbase', 'price']]

print(stats.pearsonr(df1_2['wheelbase'], df1_2['price']))

PearsonRResult(statistic=0.5778155982921502, pvalue=1.1828204863376843e-19)


### nhận xét: pvalue=1.1828204863376843e-19 < 0.05
--> bác bỏ H0 -> tức là 2 thuộc tính này (wheelbase và price) có tương quan

In [24]:
# carlength và price 

df1_3 = df1[['carlength', 'price']]

print(stats.pearsonr(df1_3['carlength'], df1_3['price']))

PearsonRResult(statistic=0.682920015677962, pvalue=1.6787066240746824e-29)


### nhận xét: pvalue=1.6787066240746824e-29 < 0.05
--> bác bỏ H0 -> tức là 2 thuộc tính này (carlength và price) có tương quan

In [25]:
# carwidth và price 

df1_4 = df1[['carwidth', 'price']]

print(stats.pearsonr(df1_4['carwidth'], df1_4['price']))

PearsonRResult(statistic=0.7593252997415115, pvalue=9.627437876119045e-40)


### nhận xét: pvalue=9.627437876119045e-40 < 0.05
--> bác bỏ H0 -> tức là 2 thuộc tính này (carwidth và price) có tương quan

In [26]:
# curbweight và price  
df1_5 = df1[['curbweight', 'price']]

print(stats.pearsonr(df1_5['curbweight'], df1_5['price']))

PearsonRResult(statistic=0.8353048793372966, pvalue=1.2144448264032194e-54)


### nhận xét: pvalue=1.2144448264032194e-54 < 0.05
--> bác bỏ H0 -> tức là 2 thuộc tính này (curbweight và price) có tương quan

In [27]:
# enginesize và price 

df1_6 = df1[['enginesize', 'price']]

print(stats.pearsonr(df1_6['enginesize'], df1_6['price']))

PearsonRResult(statistic=0.8741448025245117, pvalue=1.3547637598649229e-65)


### nhận xét: pvalue=1.3547637598649229e-65 < 0.05
--> bác bỏ H0 -> tức là 2 thuộc tính này (enginesize và price) có tương quan

In [28]:
# boreratio và price  
df1_7 = df1[['boreratio', 'price']]

print(stats.pearsonr(df1_7['boreratio'], df1_7['price']))

PearsonRResult(statistic=0.5531732367984437, pvalue=7.90792227824885e-18)


### nhận xét: pvalue=7.90792227824885e-18 < 0.05
--> bác bỏ H0 -> tức là 2 thuộc tính này (boreratio và price) có tương quan

In [29]:
# highwaympg và price 
df1_8 = df1[['highwaympg', 'price']]

print(stats.pearsonr(df1_8['highwaympg'], df1_8['price']))

PearsonRResult(statistic=-0.6975990916465561, pvalue=3.230681380364934e-31)


### nhận xét: pvalue=3.230681380364934e-31 < 0.05
--> bác bỏ H0 -> tức là 2 thuộc tính này (highwaympg và price) có tương quan

## Nhận xét: 

Sau khi chạy kiếm định tương quan các biến cylindernumber, wheelbase,carlength, carwidth, curbweight, enginesize, boreratio, highwaympg với giá (price)

thì thấy được rằng các biến đều có mối tương quan với giá. 
Các giá p-values >0 cho nên có thể thấy được đây là tương quan thuận

## Vẽ biểu đồ heatmap biểu diễn mức độ tương quan giữa các biến định lượng (xem xét có những biến nào có tương quan lớn với nhau, chỉ giữa lại 1 trong các biến đó đưa vào mô hình

## Tiến hành mã hóa tạo ra các biến dumy từ các biến category (thang đo nominal) ban đầu

In [32]:
# tiến hành dummy các biến bảng df2 (trừ biến price)

dummy = pd.get_dummies(df2.drop(['price'], axis=1)).astype(int)
dummy

CarName_Nissan versa  CarName_alfa-romero Quadrifoglio  \
0                       0                                 0   
1                       0                                 0   
2                       0                                 1   
3                       0                                 0   
4                       0                                 0   
..                    ...                               ...   
200                     0                                 0   
201                     0                                 0   
202                     0                                 0   
203                     0                                 0   
204                     0                                 0   

     CarName_alfa-romero giulia  CarName_alfa-romero stelvio  \
0                             1                            0   
1                             0                            1   
2                             0                            0   
3                             0                            0   
4                             0                            0   
..                          ...                          ...   
200                           0                            0   
201                           0                            0   
202                           0                            0   
203                           0                            0   
204                           0                            0   

     CarName_audi 100 ls  CarName_audi 100ls  CarName_audi 4000  \
0                      0                   0                  0   
1                      0                   0                  0   
2                      0                   0                  0   
3                      1                   0                  0   
4                      0                   1                  0   
..                   ...                 ...                ...   
200                    0                   0                  0   
201                    0                   0                  0   
202                    0                   0                  0   
203                    0                   0                  0   
204                    0                   0                  0   

     CarName_audi 5000  CarName_audi 5000s (diesel)  CarName_audi fox  ...  \
0                    0                            0                 0  ...   
1                    0                            0                 0  ...   
2                    0                            0                 0  ...   
3                    0                            0                 0  ...   
4                    0                            0                 0  ...   
..                 ...                          ...               ...  ...   
200                  0                            0                 0  ...   
201                  0                            0                 0  ...   
202                  0                            0                 0  ...   
203                  0                            0                 0  ...   
204                  0                            0                 0  ...   

     enginetype_ohcv  enginetype_rotor  fuelsystem_1bbl  fuelsystem_2bbl  \
0                  0                 0                0                0   
1                  0                 0                0                0   
2                  1                 0                0                0   
3                  0                 0                0                0   
4                  0                 0                0                0   
..               ...               ...              ...              ...   
200                0                 0                0                0   
201                0                 0                0                0   
202                1                 0               

## Xây dựng mô hình hồi quy tuyến tính bằng thư viện sklearn, sử dụng thêm feature selection để chọn các thuộc tính tốt nhất.

In [36]:
# nối bảng dummy với cột price

dummy1 = pd.concat([dummy, df2[['price']]], axis=1)
dummy1

CarName_Nissan versa  CarName_alfa-romero Quadrifoglio  \
0                       0                                 0   
1                       0                                 0   
2                       0                                 1   
3                       0                                 0   
4                       0                                 0   
..                    ...                               ...   
200                     0                                 0   
201                     0                                 0   
202                     0                                 0   
203                     0                                 0   
204                     0                                 0   

     CarName_alfa-romero giulia  CarName_alfa-romero stelvio  \
0                             1                            0   
1                             0                            1   
2                             0                            0   
3                             0                            0   
4                             0                            0   
..                          ...                          ...   
200                           0                            0   
201                           0                            0   
202                           0                            0   
203                           0                            0   
204                           0                            0   

     CarName_audi 100 ls  CarName_audi 100ls  CarName_audi 4000  \
0                      0                   0                  0   
1                      0                   0                  0   
2                      0                   0                  0   
3                      1                   0                  0   
4                      0                   1                  0   
..                   ...                 ...                ...   
200                    0                   0                  0   
201                    0                   0                  0   
202                    0                   0                  0   
203                    0                   0                  0   
204                    0                   0                  0   

     CarName_audi 5000  CarName_audi 5000s (diesel)  CarName_audi fox  ...  \
0                    0                            0                 0  ...   
1                    0                            0                 0  ...   
2                    0                            0                 0  ...   
3                    0                            0                 0  ...   
4                    0                            0                 0  ...   
..                 ...                          ...               ...  ...   
200                  0                            0                 0  ...   
201                  0                            0                 0  ...   
202                  0                            0                 0  ...   
203                  0                            0                 0  ...   
204                  0                            0                 0  ...   

     enginetype_rotor  fuelsystem_1bbl  fuelsystem_2bbl  fuelsystem_4bbl  \
0                   0                0                0                0   
1                   0                0                0                0   
2                   0                0                0                0   
3                   0                0                0                0   
4                   0                0                0                0   
..                ...              ...              ...              ...   
200                 0                0                0                0   
201                 0                0                0                0   
202                 0                0               

In [38]:
# tìm thuộc tính có mối quan hệ với biến price ở bảng dummy1

d = dummy1.corr()
d

CarName_Nissan versa  \
CarName_Nissan versa                          1.000000   
CarName_alfa-romero Quadrifoglio             -0.004902   
CarName_alfa-romero giulia                   -0.004902   
CarName_alfa-romero stelvio                  -0.004902   
CarName_audi 100 ls                          -0.004902   
...                                                ...   
fuelsystem_mfi                               -0.004902   
fuelsystem_mpfi                              -0.064430   
fuelsystem_spdi                              -0.015003   
fuelsystem_spfi                              -0.004902   
price                                        -0.068330   

                                  CarName_alfa-romero Quadrifoglio  \
CarName_Nissan versa                                     -0.004902   
CarName_alfa-romero Quadrifoglio                          1.000000   
CarName_alfa-romero giulia                               -0.004902   
CarName_alfa-romero stelvio                              -0.004902   
CarName_audi 100 ls                                      -0.004902   
...                                                            ...   
fuelsystem_mfi                                           -0.004902   
fuelsystem_mpfi                                           0.076082   
fuelsystem_spdi                                          -0.015003   
fuelsystem_spfi                                          -0.004902   
price                                                     0.028318   

                                  CarName_alfa-romero giulia  \
CarName_Nissan versa                               -0.004902   
CarName_alfa-romero Quadrifoglio                   -0.004902   
CarName_alfa-romero giulia                          1.000000   
CarName_alfa-romero stelvio                        -0.004902   
CarName_audi 100 ls                                -0.004902   
...                                                      ...   
fuelsystem_mfi                                     -0.004902   
fuelsystem_mpfi                                     0.076082   
fuelsystem_spdi                                    -0.015003   
fuelsystem_spfi                                    -0.004902   
price                                               0.001918   

                                  CarName_alfa-romero stelvio  \
CarName_Nissan versa                                -0.004902   
CarName_alfa-romero Quadrifoglio                    -0.004902   
CarName_alfa-romero giulia                          -0.004902   
CarName_alfa-romero stelvio                          1.000000   
CarName_audi 100 ls                                 -0.004902   
...                                                       ...   
fuelsystem_mfi                                      -0.004902   
fuelsystem_mpfi                                      0.076082   
fuelsystem_spdi                                     -0.015003   
fuelsystem_spfi                                     -0.004902   
price                                                0.028318   

                                  CarName_audi 100 ls  CarName_audi 100ls  \
CarName_Nissan versa                        -0.004902           -0.006949   
CarName_alfa-romero Quadrifoglio            -0.004902           -0.006949   
CarName_alfa-romero giulia                  -0.004902           -0.006949   
CarName_alfa-romero stelvio                 -0.004902           -0.006949   
CarName_audi 100 ls                          1.000000           -0.006949   
...                                               ...                 ...   
fuelsystem_mfi                              -0.004902           -0.006949   
fuelsystem_mpfi                              0.076082            0.107861   
fuelsystem_spdi                             -0.015003           -0.021270   
fuelsystem_spfi                             -0.004902           -0.006949   
price                                        0.005915            0.053598   

                     

In [39]:
d['price'].describe()

count    177.000000
mean       0.010558
std        0.152001
min       -0.601950
25%       -0.052552
50%       -0.026582
75%        0.042357
max        1.000000
Name: price, dtype: float64

In [40]:
# Lọc ra các dòng ở bảng d trên mà có cột price >0

d1 = d[d['price'] >0]
d1

CarName_Nissan versa  \
CarName_alfa-romero Quadrifoglio             -0.004902   
CarName_alfa-romero giulia                   -0.004902   
CarName_alfa-romero stelvio                  -0.004902   
CarName_audi 100 ls                          -0.004902   
CarName_audi 100ls                           -0.006949   
...                                                ...   
enginetype_ohcf                              -0.019672   
enginetype_ohcv                              -0.018218   
fuelsystem_idi                               -0.023020   
fuelsystem_mpfi                              -0.064430   
price                                        -0.068330   

                                  CarName_alfa-romero Quadrifoglio  \
CarName_alfa-romero Quadrifoglio                          1.000000   
CarName_alfa-romero giulia                               -0.004902   
CarName_alfa-romero stelvio                              -0.004902   
CarName_audi 100 ls                                      -0.004902   
CarName_audi 100ls                                       -0.006949   
...                                                            ...   
enginetype_ohcf                                          -0.019672   
enginetype_ohcv                                           0.269069   
fuelsystem_idi                                           -0.023020   
fuelsystem_mpfi                                           0.076082   
price                                                     0.028318   

                                  CarName_alfa-romero giulia  \
CarName_alfa-romero Quadrifoglio                   -0.004902   
CarName_alfa-romero giulia                          1.000000   
CarName_alfa-romero stelvio                        -0.004902   
CarName_audi 100 ls                                -0.004902   
CarName_audi 100ls                                 -0.006949   
...                                                      ...   
enginetype_ohcf                                    -0.019672   
enginetype_ohcv                                    -0.018218   
fuelsystem_idi                                     -0.023020   
fuelsystem_mpfi                                     0.076082   
price                                               0.001918   

                                  CarName_alfa-romero stelvio  \
CarName_alfa-romero Quadrifoglio                    -0.004902   
CarName_alfa-romero giulia                          -0.004902   
CarName_alfa-romero stelvio                          1.000000   
CarName_audi 100 ls                                 -0.004902   
CarName_audi 100ls                                  -0.006949   
...                                                       ...   
enginetype_ohcf                                     -0.019672   
enginetype_ohcv                                     -0.018218   
fuelsystem_idi                                      -0.023020   
fuelsystem_mpfi                                      0.076082   
price                                                0.028318   

                                  CarName_audi 100 ls  CarName_audi 100ls  \
CarName_alfa-romero Quadrifoglio            -0.004902           -0.006949   
CarName_alfa-romero giulia                  -0.004902           -0.006949   
CarName_alfa-romero stelvio                 -0.004902           -0.006949   
CarName_audi 100 ls                          1.000000           -0.006949   
CarName_audi 100ls                          -0.006949            1.000000   
...                                               ...                 ...   
enginetype_ohcf                             -0.019672           -0.027889   
enginetype_ohcv                             -0.018218           -0.025828   
fuelsystem_idi                              -0.023020           -0.032636   
fuelsystem_mpfi                              0.076082            0.107861   
price                                        0.005915            0.053598   

                     

In [41]:
# Lọc ra các dòng ở bảng d trên mà có cột price có giá trị >0.3

d2 = d[d['price'] >0.3]
d2

CarName_Nissan versa  CarName_alfa-romero Quadrifoglio  \
drivewheel_rwd                  -0.053740                          0.091216   
enginelocation_rear             -0.008532                         -0.008532   
enginetype_ohcv                 -0.018218                          0.269069   
fuelsystem_mpfi                 -0.064430                          0.076082   
price                           -0.068330                          0.028318   

                     CarName_alfa-romero giulia  CarName_alfa-romero stelvio  \
drivewheel_rwd                         0.091216                     0.091216   
enginelocation_rear                   -0.008532                    -0.008532   
enginetype_ohcv                       -0.018218                    -0.018218   
fuelsystem_mpfi                        0.076082                     0.076082   
price                                  0.001918                     0.028318   

                     CarName_audi 100 ls  CarName_audi 100ls  \
drivewheel_rwd                 -0.053740           -0.076187   
enginelocation_rear            -0.008532           -0.012096   
enginetype_ohcv                -0.018218           -0.025828   
fuelsystem_mpfi                 0.076082            0.107861   
price                           0.005915            0.053598   

                     CarName_audi 4000  CarName_audi 5000  \
drivewheel_rwd               -0.053740          -0.053740   
enginelocation_rear          -0.008532          -0.008532   
enginetype_ohcv              -0.018218          -0.018218   
fuelsystem_mpfi               0.076082           0.076082   
price                         0.093110           0.049579   

                     CarName_audi 5000s (diesel)  CarName_audi fox  ...  \
drivewheel_rwd                         -0.053740         -0.053740  ...   
enginelocation_rear                    -0.008532         -0.008532  ...   
enginetype_ohcv                        -0.018218         -0.018218  ...   
fuelsystem_mpfi                         0.076082          0.076082  ...   
price                                   0.040259          0.017336  ...   

                     enginetype_rotor  fuelsystem_1bbl  fuelsystem_2bbl  \
drivewheel_rwd               0.183789        -0.182771        -0.464056   
enginelocation_rear         -0.017192        -0.029019        -0.083975   
enginetype_ohcv             -0.036707        -0.061961        -0.179302   
fuelsystem_mpfi             -0.059039        -0.219128        -0.634114   
price                       -0.004544        -0.170945        -0.501374   

                     fuelsystem_4bbl  fuelsystem_idi  fuelsystem_mfi  \
drivewheel_rwd              0.158772        0.122035       -0.053740   
enginelocation_rear        -0.014851       -0.040070       -0.008532   
enginetype_ohcv            -0.031711       -0.085556       -0.018218   
fuelsystem_mpfi            -0.112147       -0.302574       -0.064430   
price                      -0.017306        0.105679       -0.002747   

                     fuelsystem_mpfi  fuelsystem_spdi  fuelsystem_spfi  \
drivewheel_rwd              0.448977        -0.115182         0.091216   
enginelocation_rear         0.132429        -0.026114        -0.008532   
enginetype_ohcv             0.282760        -0.055759        -0.018218   
fuelsystem_mpfi             1.000000        -0.197195        -0.064430   
price                       0.517075        -0.061475        -0.019580   

                        price  
drivewheel_rwd       0.638957  
enginelocation_rear  0.324973  
enginetype_ohcv      0.385991  
fuelsystem_mpfi      0.517075  
price                1.000000  

[5 rows x 177 columns]

### Như vậy có thể thấy được chỉ có 4 biến category có mối quan hệ với giá (price)

drivewheel_rwd, enginelocation_rear, enginetype_ohcv, fuelsystem_mpfi	

In [43]:
# mình sẽ nối 4 cột biến này ở bảng dummy với 8 cột biến định lượng ở trên bảng df1 mình đã phân tích có tương quan với giá 
# đế tiến hành chạy mô hình hồi quy tuyến tính

df3 = pd.concat([dummy[['drivewheel_rwd', 'enginelocation_rear', 'enginetype_ohcv', 'fuelsystem_mpfi']],
                df1[['cylindernumber', 'wheelbase', 'carlength', 'carwidth', 'curbweight', 'enginesize', 'boreratio', 'highwaympg', 'price']]], 
                 axis=1)
df3

drivewheel_rwd  enginelocation_rear  enginetype_ohcv  fuelsystem_mpfi  \
0                 1                    0                0                1   
1                 1                    0                0                1   
2                 1                    0                1                1   
3                 0                    0                0                1   
4                 0                    0                0                1   
..              ...                  ...              ...              ...   
200               1                    0                0                1   
201               1                    0                0                1   
202               1                    0                1                1   
203               1                    0                0                0   
204               1                    0                0                1   

     cylindernumber  wheelbase  carlength  carwidth  curbweight  enginesize  \
0                 4       88.6      168.8      64.1        2548         130   
1                 4       88.6      168.8      64.1        2548         130   
2                 6       94.5      171.2      65.5        2823         152   
3                 4       99.8      176.6      66.2        2337         109   
4                 5       99.4      176.6      66.4        2824         136   
..              ...        ...        ...       ...         ...         ...   
200               4      109.1      188.8      68.9        2952         141   
201               4      109.1      188.8      68.8        3049         141   
202               6      109.1      188.8      68.9        3012         173   
203               6      109.1      188.8      68.9        3217         145   
204               4      109.1      188.8      68.9        3062         141   

     boreratio  highwaympg    price  
0         3.47          27  13495.0  
1         3.47          27  16500.0  
2         2.68          26  16500.0  
3         3.19          30  13950.0  
4         3.19          22  17450.0  
..         ...         ...      ...  
200       3.78          28  16845.0  
201       3.78          25  19045.0  
202       3.58          23  21485.0  
203       3.01          27  22470.0  
204       3.78          25  22625.0  

[205 rows x 13 columns]

##### Chạy mô hình hồi quy tuyến tính trên bảng df3 

In [44]:
# Chia ngẫu nhiên dữ liệu quan sát được thành hai phần dữ liệu huấn luyện và dữ liệu kiểm tra.

X = df3.drop(['price'], axis=1)
y = df3['price']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 102)


In [45]:
# Xây dựng mô hình hồi quy tuyến tính đa biến với biến dự báo là giá, các biến đầu vào là các trường còn lại.

lr = LinearRegression()
lr.fit(X_train, y_train)

LinearRegression()

In [48]:
# Đánh giá mô hình.
from sklearn import metrics
y_pred = lr.predict(X_test)
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 1820.2970199374463
Mean Squared Error: 6247648.432353775
Root Mean Squared Error: 2499.5296422234674


In [47]:
df3['price'].describe()

count      205.000000
mean     13276.710571
std       7988.852332
min       5118.000000
25%       7788.000000
50%      10295.000000
75%      16503.000000
max      45400.000000
Name: price, dtype: float64

#### Giải thích các chỉ số trên:
Mean Absolute Error = 1820.297: trung bình giá dự đoán đang lệch khỏi giá trung bình thật 1820 (đv tiền tệ)

Root Mean Squared Error = 2499: nếu đánh giá kết quả chặt chẽ hơn, thì trung bình giá dự đoán đang lệch khỏi giá trung bình thật 2499 (đv tiền tệ)

giá trung bình là 13276, đọ lệch chuẩn 7988--> giá trị giá lệch  là thấp < std, do vậy mô hình này 
chấp nhận được

In [49]:
# Kiểm tra thêm giá trị  r2_score

score= r2_score(y_test,y_pred)
score

0.8342296530142235

#### Nhận xét thêm:

chỉ số này cho biết thêm mối quan hệ giữa các biến trục X và y.

chỉ số này  = 0.834 tương đương 83.4%, có thể thấy được mô hình khá tốt và phù hợp



## Xây dựng mô hình hồi quy tuyến tính trên các biến được chọn bởi phương pháp feature selection. Có thể sử dụng thêm các thư viện khác để xây dựng mô hình như statsAPI.OLS

#### Đánh giá chất lượng 2 mô hình dựa trên các thông số như R_square, Tổng bình phương sai số trung bình.